terraform/manifest/main.tf

```diff
@@ -13,8 +13,36 @@ locals {
     units       = null
     storage     = null
   }
+
+  _allowed_exposed = ["cidrs", "endpoints", "spaces"]
+  exposed = {
+    for app, obj in local.full_map : app => {
+      expose = (
+        # if the expose data is not present, return null
+        !contains(keys(obj), "expose") || obj.expose == null ? null :
+
+        # if the expose data is utterly empty, return null
+        length([
+          for k in local._allowed_exposed :
+          k if contains(keys(obj.expose), k) && obj.expose[k] != null
+        ]) == 0 ? null :
+
+        # Filter the expose data to only include allowed keys
+        # and ensure that no null values are included
+        {
+          for k, v in obj.expose : k => v
+          if contains(local._allowed_exposed, k) && v != null
+        }
+      )
+    }
+  }
```

Comment:
builds the juju `exposed` map for all the applications in the manifest.yaml

* nulls out the object if its empty, missing, or has only empty keys
* filters out any unexpected keys